onnxruntime/core/providers/cuda/math/matmul_integer.cc

```diff
@@ -0,0 +1,156 @@
+    auto a_zero_point = ctx->Input<Tensor>(2);
+    ORT_ENFORCE(IsScalarOr1ElementVector(a_zero_point),
+                "MatmulInteger : input1 zero point must be a scalar or 1D tensor of size 1");
+    a_offset = a_zero_point->template Data<int8_t>();
+  }
+  if (has_b_zero_point_) {
+    auto b_zero_point = ctx->Input<Tensor>(3);
+    ORT_ENFORCE(IsScalarOr1ElementVector(b_zero_point),
+                "MatmulInteger : input2 zero point must be a scalar or 1D tensor of size 1");
+    b_offset = b_zero_point->template Data<int8_t>();
+  }
+
+  // offset output c[i,j] to
+  // k*a_offset*b_offset -
+  // b_offset * (a[i,0] + a[i,1] ...+a[i,k]) -
+  // a_offset * (b[0,j] + b[1,j] ... + b[k,j])
+  // ReduceRowSumOnMatrixA computes the b_offset * (a[i,0] + a[i,1] ...+a[i,k]) part
+  // ReduceColSumOnMatrixB computes the a_offset * (b[0,j] + b[1,j] ... + b[k,j]) part
+  // OffsetOutput computes gets the final result
+  IAllocatorUniquePtr<int32_t> a_row_buf;
+  if (has_b_zero_point_) {
+    a_row_buf = GetScratchBuffer<int32_t>(helper.OutputShape().Size() / helper.N());
+    ORT_RETURN_IF_ERROR(ReduceRowSumOnMatrixA(a_ptr, a_row_buf.get(), b_offset, helper));
+  }
+
+  IAllocatorUniquePtr<int32_t> b_col_buf;
+  if (has_a_zero_point_) {
+    b_col_buf = GetScratchBuffer<int32_t>(helper.OutputShape().Size() / helper.M());
+    ORT_RETURN_IF_ERROR(ReduceColSumOnMatrixB(b_ptr, b_col_buf.get(), a_offset, helper));
+  }
+
+  int alpha = 1;
+  int beta = 0;
+  if (has_a_zero_point_ || has_b_zero_point_) {
+    OffsetOutput(a_row_buf.get(),
+                 b_col_buf.get(),
+                 output_ptr,
+                 a_offset,
+                 b_offset,
+                 helper);
+    beta = 1;
+  }
+
+  // pad A and B to make their leading dimension be multiples of 32
+  // because cublasGemmEx requires:
+  // 1. leading dimension is multiples of 4
+  // 2. A, B is 32-bit aligned
+  const int64_t align_size = 32;
+  int64_t a_pad_size = 0;
+  int64_t b_pad_size = 0;
+  IAllocatorUniquePtr<int8_t> a_padded;
+  IAllocatorUniquePtr<int8_t> b_padded;
+  ORT_RETURN_IF_ERROR(PadMatrix(a->Shape().Size() / helper.K(),
+                                helper.K(),
+                                align_size,
+                                a_ptr,
+                                a_pad_size,
+                                a_padded));
+  ORT_RETURN_IF_ERROR(PadMatrix(b->Shape().Size() / helper.N(),
+                                helper.N(),
+                                align_size,
+                                b_ptr,
+                                b_pad_size,
+                                b_padded));
+
+  for (int batch = 0; batch < helper.OutputOffsets().size(); batch++) {
+    CUBLAS_RETURN_IF_ERROR(cublasGemmEx(
+        Base::CublasHandle(),
+        CUBLAS_OP_N,
+        CUBLAS_OP_N,
+        static_cast<int>(helper.N()),
+        static_cast<int>(helper.M()),
+        static_cast<int>(helper.K()),
+        &alpha,
```

Comment:
Any chance this conversion from int64_t to int would ever overflow and should be checked?

Reply:
a matrix with a dimension > 2^31 is extremely rare in practice.